pageserver/src/tenant/layer_map/historic_layer_coverage.rs

```diff
@@ -204,6 +204,31 @@ fn test_off_by_one() {
     assert_eq!(version.image_coverage.query(5), None);
 }
 
+/// White-box regression test, checking for incorrect removal of node at key.end
+#[test]
+fn test_regression() {
+    let mut map = HistoricLayerCoverage::<String>::new();
+    map.insert(
+        LayerKey {
+            key: 0..5,
+            lsn: 0..5,
+            is_image: false,
+        },
+        "Layer 1".to_string(),
+    );
+    map.insert(
+        LayerKey {
+            key: 0..5,
+            lsn: 1..2,
+            is_image: false,
+        },
+        "Layer 2".to_string(),
+    );
+
+    let version = map.get_version(100).unwrap();
+    assert_eq!(version.delta_coverage.query(100), None);
```

Comment:
For me this was quite difficult to see how this will fail without trying:

```
---- tenant::layer_map::historic_layer_coverage::test_regression stdout ----
thread 'tenant::layer_map::historic_layer_coverage::test_regression' panicked at 'assertion failed: `(left == right)`
  left: `Some("Layer 1")`,
 right: `None`', pageserver/src/tenant/layer_map/historic_layer_coverage.rs:831:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Would prefer to see

```suggestion
    assert_eq!(
        version.delta_coverage.query(100),
        None,
        "neither layer covers the key"
    );
```

Above suggestion does not really explain the effect this had and I am not really sure what would had happened either.

It could also be expanded as:

```suggestion
    assert_eq!(
        version.delta_coverage.query(100),
        None,
        "neither layer covers the key"
    );
    
        for key in 0..5 {
        assert_eq!(
            version.delta_coverage.query(key).as_deref(),
            Some("Layer 1"),
            "key {key} is covered by the nearest covering layer"
        );
    }
```

But perhaps the second suggestion is just me having hard time getting up to speed.

Reply:
It's a white-box test, so the explanation is "some past implementation got this wrong, trust me". But we can try to explain with a comment "If an insertion operation improperly deletes the endpoint of a previous layer (which is more likely to happen with layers that collide on key.end), we will end up with an infinite layer, covering the entire keyspace. Here we assert that there's no layer at key 100 because we didn't insert any layer there."

Reply:
Oki, I am fine with it, resolving.